### PR TITLE
[web-animations-2] Make active boundary times inclusive if timeline duration is finite

### PR DESCRIPTION
**Pull request by kevers-google**
[web-animations-2] Make active boundary times inclusive if timeline duration is finite

Addresses issue #5223. The discussion suggested adding rules for resolving fill "auto".  This approach introduces edge cases as indicated in comments on https://github.com/w3c/csswg-drafts/pull/6673.  In short, resolving fill 'auto' introduces potential surprising behavior when delays or timeline offsets are non-zero.  The before and after phase calculations had explicit rules to make an active boundary exclusive depending on the direction of the animation.  This PR restricts these exclusion rules to non-finite timelines.

### CODE REVIEW

File: web-animations-2/Overview.bs [hunk 1]
```diff
@@ -1096,8 +1096,9 @@ with:
 >     <em>and</em> <em>either</em> of the following conditions are met:
 >     1.  the <a>local time</a> is less than the <a>before-active boundary
 >         time</a>, <em>or</em>
->     1.  the <a>animation direction</a> is "backwards" and the
->         <a>local time</a> is equal to the <a>before-active boundary time</a>.
+>     1.  [=timeline duration=] is unresolved, the <a>animation direction</a>
+>         is "backwards" and the <a>local time</a> is equal to the
+>         <a>before-active boundary time</a>.
```
Comment (flackr) [inline web-animations-2/Overview.bs:1101]: This behavior of making the endpoints inclusive when timeline duration is resolved is a strange special case to have. This will also introduce many of the same edge cases from #6673 even though they will be limited to the endpoint of one animation, e.g.:

```js
element.animate({transform: ['translateY(0px)', 'translateY(10px)']}, {duration: CSS.percent(10), timeline: scrollTimeline, composite: 'add'}); 
element.animate({transform: ['translateY(10px)', 'translateY(50px)']}, {duration: CSS.percent(10), startDelay: CSS.percent(10), timeline: scrollTimeline, composite: 'add'})
```

At time 10% both effects will be active resulting in a transform of 20px. The same thing will happen for back to back effects in a SequenceEffect.

Comment (flackr) [reply web-animations-2/Overview.bs:1101]: FWIW, I do like the approach of making the active phase endpoint inclusive more than the auto fill mode, it just needs to only apply when the effect reaches the end of the timeline.

Comment (flackr) [reply web-animations-2/Overview.bs:1101]: Perhaps the simplest way to address this is calculate the effective timeline time by undoing the current time calculation, e.g.
effective timeline time = (local time / playback rate) + start_time

And then only use the endpoint inclusive condition if the effective timeline time is 0 (and backwards) or 100% (and forwards)

Comment (kevers-google) [reply web-animations-2/Overview.bs:1101]: It appears that this reverse engineering of effective timeilne time is not quite right for sequence effects since local time resets to zero at the start of each effect in the sequence.  Can likely come up with an equation based on inherited time to determine if aligned with the boundary of the timeline.  Just to be clear, are we only applying if scrollOffset = [0%, 100%] (or auto) and not at an arbitrary effective start or end offset?

Comment (flackr) [reply web-animations-2/Overview.bs:1101]: local_time is zero, but the start_time would be non-zero wouldn't it? Of course there may be simpler approaches that make it less likely we'll have to deal with equality of values.

Dealing with the actual max scroll offset is a good question, I think ideally we'd only apply this for the 100% / auto case so that stringing together animations on separate scroll timelines which stop when the next one starts also doesn't result in any overlap.

Comment (kevers-google) [reply web-animations-2/Overview.bs]: Inferring the timeline time based on the inherited time and start time is only strictly correct for a running animation.  If the animation is paused we cannot tell if there was any lead/lag between the timeline and animation currentTime at the time of the pause.  Added an issue highlighting the problem. This discrepancy would only trigger an unexpected discrepancy for edge cases such as setting current time when pausing a progress-based animation precisely at the timeline boundary.  I don't believe these potential edge cases will be frequently triggered.

**Review by flackr — COMMENTED** [review]
I think this works but I'd appreciate a second set of eyes on this, @birtles would you be able to take a look?

File: web-animations-2/Overview.bs [hunk 2]
```diff
@@ -1066,13 +1066,36 @@ condition:
...
+    :   If the associated <a>animation</a>'s [=timeline duration=] is
+        unresolved,
+    ::  &rarr; unresolved
+
+    :   If the associated <a>animation</a>'s startTime is unresolved,
+    ::  &rarr; <a>inherited time</a>
+
+    :   Otherwise,
+    ::  <code><a>inherited time</a> / <a>animation</a>'s
+        playbackRate + <a>animation</a>'s startTime</code>
```
Comment (birtles) [inline web-animations-2/Overview.bs:1090]: Similarly here, for consistency it might be better to use the model terms instead of the interface member names?

**Review by birtles — COMMENTED** [review]
From what I can tell the approach seems reasonable but I have a few questions about some of the minor details.

File: web-animations-2/Overview.bs [hunk 3]
```diff
@@ -1066,13 +1066,36 @@ condition:
...
+::  Set based on the first matching condition:
+
+    <dl class="switch">
+
+    :   If the associated <a>animation</a>'s [=timeline duration=] is
+        unresolved,
+    ::  &rarr; unresolved
+
+    :   If the associated <a>animation</a>'s startTime is unresolved,
+    ::  &rarr; <a>inherited time</a>
```
[the remark this answers is not in the document]
Comment (kevers-google) [reply web-animations-2/Overview.bs]: Greatly simplified the logic here. Normalizing the time to convert to a fraction to be more consistent with other uses of the term 'progress'.

File: web-animations-2/Overview.bs (hunk 2 above)
Comment (kevers-google) [reply web-animations-2/Overview.bs]: Done.

File: web-animations-2/Overview.bs [hunk 4]
```diff
@@ -1066,13 +1066,36 @@ condition:
...
+    :   Otherwise,
+    ::  <code><a>inherited time</a> / <a>animation</a>'s
+        playbackRate + <a>animation</a>'s startTime</code>
+
+    </dl>
+
+Issue: [=effective timeline progress=] is not strictly correct for a paused
+animation if the <a>animation</a>'s currentTime is explicitly set, as this can
+introduce a lead or lag, between the <a>timeline<a>'s currentTime and
+<a>animation</a>'s currentTime.
```
[the remark this answers is not in the document]
Comment (kevers-google) [reply web-animations-2/Overview.bs]: Similar to other cases of paused scroll-driven animations going out of sync.

File: web-animations-2/Overview.bs [hunk 5]
```diff
@@ -1096,8 +1119,10 @@ with:
 >     <em>and</em> <em>either</em> of the following conditions are met:
 >     1.  the <a>local time</a> is less than the <a>before-active boundary
 >         time</a>, <em>or</em>
->     1.  the <a>animation direction</a> is "backwards" and the
->         <a>local time</a> is equal to the <a>before-active boundary time</a>.
+>     1.  The <a>animation direction</a> is "backwards", the <a>local time</a>
+>         is equal to the <a>before-active boundary time</a> and
+>         [=effective timeline progress=] is not equal to
+>         <code>CSS.percent(0)</code>.
```
[the remark this answers is not in the document]
Comment (kevers-google) [reply web-animations-2/Overview.bs]: Done. Now that effective timeline progress is normalized, it'll be unresolved, or between 0 and 1.

File: web-animations-2/Overview.bs [hunk 6]
```diff
@@ -1066,13 +1066,36 @@ condition:
...
+:   <dfn>effective timeline progress</dfn>
+::  Set based on the first matching condition:
+
+    <dl class="switch">
+
+    :   If <em>all</em> of the following conditions are true:
+        *   the associated <a>animation</a>'s [=timeline duration=] is resolved
+            and non-zero, and
+        *   the associated <a>animation</a>'s [=playback rate=] is 1, and
+        *   the assiciated <a>animation</a>'s [=start time=] is unresolved or 0
```
Comment (birtles) [inline web-animations-2/Overview.bs:1085]: nit: s/assiciated/associated/

File: web-animations-2/Overview.bs [hunk 7]
```diff
@@ -1066,13 +1066,36 @@ condition:
...
 
+:   <dfn>effective timeline progress</dfn>
+::  Set based on the first matching condition:
+
+    <dl class="switch">
+
+    :   If <em>all</em> of the following conditions are true:
+        *   the associated <a>animation</a>'s [=timeline duration=] is resolved
+            and non-zero, and
+        *   the associated <a>animation</a>'s [=playback rate=] is 1, and
```
[the remark this answers is not in the document]
Comment (kevers-google) [reply web-animations-2/Overview.bs]: Was previously trying to avoid some edge cases for paused animations, but in the process, it needlessly restricted check on running animations.  in this version, we are assuming that the timeline and animation are aligned when paused, but otherwise adjust for changes to the playback rate and/or start time.  Not sure of a better option unless storing off a "hold timeline time" instead of the "hold phase".  Having said that, getting the phase at a boundary wrong for a scroll after pause animation is likely not that critical.  Authors can always specify a fill to get the expected behavior if they expect that the scroll-linked animation will be intermittently paused.

In short, generalized "effective timeline progress" and added "at timeline boundary".  The boundary does not discriminate based on direction since the boundary's counterpart is already inclusive.  We are making both endpoints inclusive if "at timeline boundary" is set.

**Review by birtles — APPROVED** [review]
This looks really good to me. The only thing I'm not sure about is what the behavior should be when the duration of the timeline is zero. Are we sure it should always return false in that case?

File: web-animations-2/Overview.bs [hunk 8]
```diff
@@ -1066,13 +1066,41 @@ condition:
...
+    1.  Let <var>effective timeline progress</var> be
+        <code><var>effective timeline time</var> / [=timeline duration=]</code>
+
+    1.  If <var>effective timeline progress</var> is 0 or 1, return true,
+        otherwise false.
+
+Issue: This procedure is not strictly correct for a paused
+animation if the <a>animation</a>'s current time is explicitly set, as this can
+introduce a lead or lag, between the <a>timeline<a>'s current time and
+<a>animation</a>'s current time.
```
Comment (flackr) [inline web-animations-2/Overview.bs:1102]: I'm not sure I understand this issue from the note. It might be helpful to have a specific example, e.g. if a paused animation is set to time 100% the effective timeline time will be Y%.

**Review by flackr — APPROVED** [review]
Overall, this looks good to me. I think that the condition should probably be based on whether it's a progress based timeline rather than based on the animation's duration.

**Comment by birtles** [comment]
> 
> 
> I'm not sure that the use case is for a zero-duration animation. I could be convinced either way to make the boundary inclusive or exclusive. My preference is exclusive. The rationale for my preference is that a zero-duration is never active for an animation with a monotonic timeline. You'll be in the after phase if direction is forward, or before phase if direction is backwards. If we make it active, what is the progress as the scroll position changes? Are we stuck at the start or end of the animation? It seems challenging to define "expected" behavior.

Thanks, that makes sense to me.

**Comment by birtles** [comment]
> Overall, this looks good to me. I think that the condition should probably be based on whether it's a progress based timeline rather than based on the animation's duration.

I didn't quite follow this -- I don't think the PR is based on the animation's duration? Or is this just confirming the direction of the PR is correct?

**Comment by flackr** [comment]
> > Overall, this looks good to me. I think that the condition should probably be based on whether it's a progress based timeline rather than based on the animation's duration.
> 
> I didn't quite follow this -- I don't think the PR is based on the animation's duration? Or is this just confirming the direction of the PR is correct?

I was specifically referring to this part:

> If <em>either</em> of the following conditions are true:
>        *   the associated <a>animation</a>'s [=timeline duration=] is
>            unresolved or zero

I think that if we have other fixed duration timelines it doesn't necessarily hold that we would want them to be endpoint inclusive.

**Comment by birtles** [comment]
> I was specifically referring to this part:
> 
> > If _either_ of the following conditions are true:
> > *   the associated animation's [=timeline duration=] is
> > unresolved or zero

Oh, I see, sorry I was looking for a reference to the animation duration, not the animation timeline duration! 😅

> I think that if we have other fixed duration timelines it doesn't necessarily hold that we would want them to be endpoint inclusive.

Yes, that's a good point. That suggests the term should probably be renamed too then, since "timeline boundary" suggests it applies to any timeline. If we're going to scope this to progress-based timelines (or whatever the current term is), I guess this definition should match that (e.g. "at progress timeline boundary" or whatever).

File: web-animations-2/Overview.bs [hunk 9]
```diff
@@ -1073,10 +1073,12 @@ an <a>animation effect</a>.
...
     the <a>animation</a> timeline's [=timeline phase=].
 
-:   <dfn>at timeline boundary</dfn>
+:   <dfn>at progress timeline boundary</dfn>
 ::  Determined using the following procedure:
 
-    1.  If <em>either</em> of the following conditions are true:
+    1.  If <em>any</em> of the following conditions are true:
+        *   the associated <an>animatoin</a> does not have a
+            [=progress-based timeline=], or
```
[the remark this answers is not in the document]
Comment (kevers-google) [reply web-animations-2/Overview.bs:1081]: Trying to avoid the awkwardness of using "associated" twice in the same sentence: once for the animation being associated with the effect, and a second time for the timeline being associated with the animation.

https://drafts.csswg.org/web-animations-2/#timelines:txt=timeline%20is%20a%20progress-based%20timeline

File: web-animations-2/Overview.bs [hunk 10]
```diff
@@ -1073,10 +1073,12 @@ an <a>animation effect</a>.
 ::  The associated <a>animation</a>'s <a>hold phase</a> if resolved, otherwise
     the <a>animation</a> timeline's [=timeline phase=].
 
-:   <dfn>at timeline boundary</dfn>
+:   <dfn>at progress timeline boundary</dfn>
 ::  Determined using the following procedure:
 
-    1.  If <em>either</em> of the following conditions are true:
+    1.  If <em>any</em> of the following conditions are true:
+        *   the associated <an>animatoin</a> does not have a
```
[the remark this answers is not in the document]
Comment (kevers-google) [reply web-animations-2/Overview.bs:1080]: done
